### PR TITLE
[2201.3.0]Sync lang.regexp module with spec

### LINKED ISSUE #38524: [Improvement]: Add split function to regexp langlib
### Description

As per https://github.com/ballerina-platform/ballerina-spec/commit/b52550f3f06f1c20c6a8dcb69f17cedad2c10e92

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #38524

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by pcnfernando** [comment]
Pulled changes from https://github.com/ballerina-platform/ballerina-lang/pull/38495. Diff will update once the said PR is merged

File: langlib/lang.regexp/src/main/ballerina/regexp.bal [hunk 1]
```diff
@@ -22,13 +22,27 @@ import ballerina/jballerina.java;
...
+# by each of the regular expression's capturing groups.
+# The member with index 0 corresponds to the entire regular expression.
+# The group with index i, where i > 1,is the i-th capturing group;
+# this will be nil if the match of the regular expression did not use
+# a match of the capturing group.
+# The capturing groups within a regular expression are ordered by the position
+# of their opening parenthesis.
 public type Groups readonly & [Span, Span?...];
 
 type SpanAsTupleType [int, int, string];
```
Comment (prakanth97) [inline langlib/lang.regexp/src/main/ballerina/regexp.bal:48]: Shall we add small description for this type as well?

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal [hunk 2]
```diff
@@ -1035,6 +1035,108 @@ function testFromStringNegative() {
...
+    regexp:Groups[]? res7 = regexp:findAllGroups(re = re `(GFG)(FGF)`, str = "GFGFGFGFGFGFGFGFGF");
+    assertTrue(res7 is regexp:Groups[]);
+    regexp:Groups[] groupsArr1 = <regexp:Groups[]>res7;
+    assertEquality(3, groupsArr1.length());
+
+    string:RegExp|error x1 = regexp:fromString(str = "AB+C*D{1,4}");
+    assertTrue(x1 is string:RegExp);
+    if (x1 is string:RegExp) {
+       assertTrue(re `AB+C*D{1,4}` == x1);
+    }
```
Comment (prakanth97) [inline langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1137]: Shall we add a test for split function with namedArgExpr?

File: langlib/lang.regexp/src/main/ballerina/regexp.bal (hunk 1 above)
Comment (pcnfernando) [reply langlib/lang.regexp/src/main/ballerina/regexp.bal:48]: Those types are internal types we maintain to map between the public types. IMO, the name is self-explanatory

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal (hunk 2 above)
Comment (pcnfernando) [reply langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1137]: namedArgExpr is not related to regexp langlib. Hence I don't think namedArgExpr should be tested with each function

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal [hunk 3]
```diff
@@ -1035,6 +1035,108 @@ function testFromStringNegative() {
...
+    string[] arrExpected5 = ["ballerina", "geeks", "wso2"];
+    string[] resArr5 = re `@`.split(str5);
+    assertEquality(3, resArr5.length());
+    assertEquality(arrExpected5, resArr5);
+
+    string str6 = "ballerina.geeks.wso2";
+    string[] arrExpected6 = [str6];
+    string[] resArr6 = re `.`.split(str6);
+    assertEquality(1, resArr6.length());
+    assertEquality(arrExpected6, resArr6);
```
[the remark this answers is not in the document]
Comment (pcnfernando) [reply langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1073]: eg:
```ballerina
    string[] result = re `[a-c]`.split("zzzzayyyybxxxxawwww"); //output  ["zzzz", "yyyy", "xxxx", "wwww"]
```

[the remark this answers is not in the document]
Comment (pcnfernando) [reply langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1073]: Added in 67d5328311b199e453bad4d93c63a194ca7f5eaf

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Split.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package org.ballerinalang.langlib.regexp;
```
Comment (gimantha) [inline langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Split.java:1]: Lets add copyright notes here

File: langlib/lang.string/src/main/ballerina/string.bal [hunk 5]
```diff
@@ -311,13 +309,27 @@ public isolated function padZero(string str, int len, Char zeroChar = "0") retur
...
-# True if there is a match for `re` anywhere in `str`
-public function includesMatch(string str, RegExp 're, int startIndex = 0) returns boolean {
-   return 're.find(str, startIndex) != ();
+# Tests whether there is a match of a regular expression somewhere within a string.
+# This is equivalent to `regexp:find(re, str, startIndex) != ()`.
+#
+# + str - the string to be matched
+# + re - the regular expression
+# + return - true if the is a match of `re` somewhere within `str`, otherwise false
+public function includesMatch(string str, RegExp re, int startIndex = 0) returns boolean {
```
Comment (prakanth97) [inline langlib/lang.string/src/main/ballerina/string.bal:333]: Do we have a test for this? if not, shall we add a test?

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Split.java (hunk 4 above)
Comment (pcnfernando) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Split.java:19]: Added

File: langlib/lang.string/src/main/ballerina/string.bal (hunk 5 above)
Comment (gimantha) [reply langlib/lang.string/src/main/ballerina/string.bal:333]: Added here: https://github.com/ballerina-platform/ballerina-lang/pull/38525/commits/9973c801a9fe4550bb451aca5edb011ff2075e92

**Review by gimantha — DISMISSED** [review]

**Review by gimantha — APPROVED** [review]

**Pull request merged**